cleanup_utils/mysql_cleanup.py

```diff
@@ -0,0 +1,136 @@
+from urllib.parse import urlparse
+
+import MySQLdb as mysql
+from MySQLdb import Error as MySQLError
+
+
+# Logging ####################################################################
+
+logger = logging.getLogger('integration_cleanup')
+
+
+# Functions ##################################################################
+
+def _get_cursor(url):
+    """
+    Returns a cursor on the database
+    """
+    database_url_obj = urlparse(url)
+    try:
+        connection = mysql.connect(
+            host=database_url_obj.hostname,
+            user=database_url_obj.username or '',
+            passwd=database_url_obj.password or '',
+            port=database_url_obj.port or 3306,
+        )
+    except MySQLError as exc:
+        logger.exception('Cannot get MySQL cursor: %s', exc)
+        raise
+    return connection.cursor()
+
+
+# Classes ####################################################################
+
+class MySqlCleanupInstance:
+    """
+    Handles the cleanup of old MySQL databases
+    """
+    def __init__(self, age_limit, url, domain, dry_run):
+        """
+        Set up variables needed for cleanup
+        """
+        self.age_limit = age_limit
+        self.cleaned_up_hashes = []
+        self.domain_suffix = domain.replace('.', '_')
+        self.dry_run = dry_run
+        try:
+            self.cursor = _get_cursor(url)
+        except Exception:  # pylint: disable=broad-except
+            # Exception has already been logged
+            self.cursor = None
+
+    def _get_old_databases(self):
+        """
+        Returns a list of databases older than the age limit
+        """
+        query = """SELECT table_schema, MAX(create_time) AS create_time
+            FROM information_schema.tables
+            WHERE create_time <= DATE_SUB(NOW(), INTERVAL %(age_limit)s DAY)
+            AND table_schema LIKE %(domain_filter)s
+            GROUP BY table_schema ORDER BY create_time DESC"""
+        params = {
+            'age_limit': self.age_limit,
+            'domain_filter': '%_{}_%'.format(self.domain_suffix)
+        }
+        self.cursor.execute(query, params)
+        return self.cursor.fetchall()
+
+    def run_cleanup(self):
+        """
+        Runs the cleanup of MySQL databases older than the age limit
+        """
+        logger.info("\n --- Starting MySQL Cleanup ---")
+        if not self.cursor:
+            logger.error(
+                'ERROR: Not connected to the database, '
+                'MySQL cleanup will be skipped.'
+            )
+            return
+
+        if self.dry_run:
+            logger.info("Running in DRY_RUN mode, no actions will be taken.")
+
+        databases = self._get_old_databases()
+        logger.info('Found %d old databases', len(databases))
+
+        for (database, create_date) in databases:
+            logger.info('  > Considering database %s', database)
+            instance_database_re = r'^([0-9a-f]{8})([_0-9a-z]+)?_%s' % (self.domain_suffix,)
```

Comment:
@jamestait, this may not be required because the databases are already filtered using the base domain name and it is unlikely to have a database on this RDS server which matches the domain name but doesn't match the pattern. All the databases in this MySQL server are used solely for the CI checks.

Reply:
This is serving a dual purpose: it's an additional precaution against dropping unrelated databases, but it's also extracting the hash to return to the main script so it can be used to clean up old DNS entries. We could, perhaps, go ahead with the removal even if the RE doesn't match, in which case no hash would be returned for that database, but I've erred on the side of caution in case this is used elsewhere, where our assumptions don't hold.

Reply:
@jamestait, it will be good to mention the gist of this explanation in a comment before this block of code.

Reply:
You read my mind - added. :)